### ISSUE TITLE
Faster loading scripts?

### ISSUE THREAD

**Issue by MatthewFrench** [oleg-shilo/cs-script#34]
LoadFiles is plenty fast but Load, CreateFunc, LoadMethod, LoadScript all seem very slow.

Wouldn't avoiding creating temporary files and also avoid creating an assembly file increase speed by a lot?

**Comment by oleg-shilo** [comment]
As a starting point let's clear out a few things. 

1. Any performance penalty caused by the temp file IO is in order of magnitude (~100 times) smaller than the compilation time + startup time introduced by the actual code compiler (MS csc.exe). 
2. Creating temporary files is a natural part of a C# compilation and it cannot be avoided. 
Even if one uses `ICodeCompiler.CompileAssemblyFromSource` to compile assembly from the memory string the compiler (csc.exe) **always** creates the source and the assembly files. 
3. `CreateFunc` requires CS-Script creating one extra wrapper (auto-class) file. However this does not impact performance for the reason #1
4. Even if avoiding temp files was possible it wouldn't improve overall performance (#1). However avoiding unnecessary compilation would. This is what the caching mode is. CS-Script caching is conceptually similar to the Python caching.  
If the hosting use-case is highly compatible with caching (frequently executed non-altered scripts) then caching can be the answer that can deliver performance of a fully compiled C# code.
5. If avoiding temp files is beneficial for non-performance reasons it can be achieved by using CS-Script Evaluators model (Mono and Roslyn). Though there will be maintenance penalty for that.  Roslyn also introduces a huge startup penalty,

----

Anyway, while CS-Script is serious about performance, of course some hiccups can happen. Thus I have created a test case for this issue item:   https://github.com/oleg-shilo/cs-script/commit/e6ec82fc50986d343e3b7b8c2aae19a53f745803

Though it (as I expected) showed completely different dynamics. `LoadFiles` is the slowest one as it always emits an "umbrella script" which cannot be cached. Other two methods are much faster (>100 times) because they utilise caching at runtime.
```
LoadFiles: 2144
LoadFile: 109
CreateFunc: 37
```
And just for the sake of experiment I have disabled caching for `CreateFunc` and its speed become comparable to ``LoadFiles` (again, as expected) while still `CreateFunc` being a faster option.
```
LoadFiles: 3287
CreateFunc: 2126   
```
---

The bottom line is. Any significant iperformance gain is only possible via some sort of caching. Either CS-Script default caching or custom user defined caching. Avoiding temp files is not possible (third-party lib restrictions) and yet ineffective anyway.

**Comment by MatthewFrench** [comment]
That makes sense. I was using LoadFiles to compile 1000 scripts while 1000 separate Load compilations were quite a bit slower. The fact that it uses another process to do the compilation makes it much more understandable.
I suppose this makes Roslyn a better choice for this situation.

**Comment by oleg-shilo** [comment]
I am afraid you are misinterpreting my previous post. Both Load and LoadFiles are triggering forking csc.exe process from within System.dll. And in my experiments (captured in the unit-test I provided url for) LoadFiles was the slowest option. That completely contradicts with your observations.

>I suppose this makes Roslyn a better choice for this situation.

Yes you can switch to Roslyn by accessing the load/compile methods via `CSScript.RoslynEvaluator` or by making it a global setting: `CSScript.EvaluatorConfig.Engine = EvaluatorEngine.Roslyn;`. Switching the same way to Mono engine may give you even a better speed than Roslyn, though it will sacrifice C#6 specific syntax. 

But the switch will give you advantage only if you are loading tons of scripts that are incompatible with caching. Otherwise caching will always be a winner.

The NuGet package comes with the intensive set of samples. You were referring so far to the CS-Script.Native API (based on CodeDOM  compiler), which is reflected in the `Scripting.native.cs` file from the NuGet package. The all CS-Script.Evaluator API samples (CodeDOM, Mono and Roslyn) are located in the `Scripting.evaluator.cs` file from the same NuGet package.

**Comment by MatthewFrench** [comment]
That is strange. When I tested LoadFiles vs a for loop on Load, LoadFiles finished substantially faster. I'll strike that down as user error.

The mono engine seems worth a try. (I have already confirmed Roslyn works well for my situation)

**Comment by oleg-shilo** [comment]
Another point....
The next release will have the updated version of Mono compiler and I just quickly retested it for C# 6. I am thrilled, it now works:
```C#
dynamic script = CSScript.Evaluator.LoadMethod(@"void Hello()
                                                 {
                                                     string w = ""World"";
                                                     Console.WriteLine($""Hello {w}!"");
                                                 }");

script.Hello();
``` 
Previous version didn't. This actually brings back Mono compiler on top in the Mono vs. Roslyn scripting battle.  

The next release is scheduled for the next week.
